pandas/core/algorithms.py

```diff
@@ -1679,8 +1680,22 @@ def map_array(
         return arr.copy()
 
     # we must convert to python types
-    values = arr.astype(object, copy=False)
+    values_as_object = arr.astype(object, copy=False)
     if na_action is None:
```

Comment:
Generally I think this section would be easier to read if you just try to assign the proper `mask` and `na_value` to variables before making one file `lib.map_infer` call at the end

Reply:
Though I'm surprised we don't have something more generic to cover this - does every function handle this on its own? Seems like a common pattern

Reply:
Ok ! I tried to separate `mask` and `na_value` initialization from `lib.map_infer` call at the end.

About using something more generic, I did some experimentations with the `_ensure_data` function but if we have a `BaseMaskedArray` int64, for example, it casts it in float64 after calling `np.asarray(values)`.

But, I will explore the code to see if I can find something useful. :-)

Reply:
I think @rhshadrach runs into this pattern in groupby a lot - he may know more

Reply:
It's not clear to me what pattern you're referring to.